lib/spack/spack/solver/concretize.lp

```diff
@@ -345,6 +345,15 @@ external_conditions_hold(Package, LocalIndex) :-
 %-----------------------------------------------------------------------------
 % Variant semantics
 %-----------------------------------------------------------------------------
+% a variant is a variant of a package if it is a variant under some condition
+% and that condition holds
+variant(Package, Variant) :- variant_condition(Package, Variant, ID),
+                             condition_holds(ID).
+
+% a variant cannot be set if it is not a variant on the package
+:- variant_set(Package, Variant),
+   not variant(Package, Variant).
```

Comment:
I'll poke with this more, but we may also need an integrity constraint to prohibit a variant to take some value if the variant condition is not met:
```
:- variant_value(Package, Variant, _), not variant(Package, Variant).
```